security-tactics improvements and additions (#8)
Roblox/creator-docs: content/en-us/scripting/security/security-tactics.md

if newPart then
	print("The server created the requested part:", newPart)
elseif newPart == false then
	print("The server denied the request. No part was created")
end
```

```lua title="Script in ServerScriptService" highlight="4, 6-7, 10-17, 19, 28"
local ReplicatedStorage = game:GetService("ReplicatedStorage")

local remoteFunction = ReplicatedStorage:WaitForChild("RemoteFunctionTest")
local t = require(ReplicatedStorage:WaitForChild("t"))

> -- Create type validator in advance to avoid unnecessary overhead
> local createPartTypeValidator = t.tuple(t.instanceIsA("Player"), t.Color3, t.Vector3)

Reviewer: ```suggestion
-- Create type validator in advance to avoid unnecessary overhead
local createPartTypeValidator = t.tuple(t.instanceIsA("Player"), t.Color3, t.Vector3)

```

if newPart then
	print("The server created the requested part:", newPart)

> elseif newPart == false then

Reviewer: Can this simply be an "else" or is this accounting for some subtle distinction between it resolving to either "nil" or "false"?

Author: In this example a simple "else" can be used, yes. The "false" return is meant to very clearly discern a failed request from the client. If the event had the chance to return "nil" (ie. retrieving some data from the server that may not exist) then for both debugging and UX purposes we'd want to be able to discern between a type checking error and a valid but empty result.

It isn't applicable to this example, so can be omitted if necessary, but I think it serves as a good general example for users. Up to you!

Reviewer: Sounds good, we'll leave it as-is for clarity.

> 	print("The server denied the request. No part was created")

Reviewer: ```suggestion
	print("The server denied the request. No part was created.")
```

end
```

> ```lua title="Script in ServerScriptService" highlight="4, 6-7, 10-17, 19, 28"

Reviewer: ```suggestion
```lua title="Script in ServerScriptService" highlight="4, 7, 12, 18, 29"
```

local ReplicatedStorage = game:GetService("ReplicatedStorage")

local remoteFunction = ReplicatedStorage:WaitForChild("RemoteFunctionTest")
local t = require(ReplicatedStorage:WaitForChild("t"))

-- Create type validator in advance to avoid unnecessary overhead
local createPartTypeValidator = t.tuple(t.instanceIsA("Player"), t.Color3, t.Vector3)

-- Create new part with the passed properties
local function createPart(player, partColor, partPosition)
	-- Type check the passed arguments
	if not createPartTypeValidator(player, partColor, partPosition) then
		-- Silently return "false" if type check fails here

> 		-- Raising an error without a cooldown can be abused by malicious cheaters
> 		-- to bog down the server. Provide client feedback instead!

Reviewer: ```suggestion
		-- Raising an error without a cooldown can be abused to bog down the server
		-- Provide client feedback instead!
```

end
```

Another common attack that exploiters may use involves sending `Library.table|tables` in place of an `Class.Instance`. Complex payloads can mimic what would be an otherwise ordinary object reference.

For example, provided with an [in-experience shop](#in-experience-shop) system where item data like prices are stored in `Class.NumberValue` objects, an exploiter may circumvent all other checks by doing the following:

> ```lua title="LocalScript in StarterPlayerScripts" highlight="5, 18, 21"
> local ReplicatedStorage = game:GetService("ReplicatedStorage")
> 
> local itemDataFolder = ReplicatedStorage:WaitForChild("ItemData")
> local buyItemEvent = ReplicatedStorage:WaitForChild("BuyItemEvent")
> local payload = {
> 	Name = "Ultra Blade",
> 	ClassName = "Folder",
> 	Parent = itemDataFolder,
> 
> 	Price = {
> 		Name = "Price",
> 		ClassName = "NumberValue",
> 		Value = 0,  -- Negative values could also be used, resulting in giving currency rather than taking it!
> 	},
> }
> 
> -- Send malicious payload to the server (this will be rejected)
> print(buyItemEvent:InvokeServer(payload))  -- Outputs "false Invalid item provided"
> 
> -- Send a real item to the server (this will go through!)
> print(buyItemEvent:InvokeServer(itemDatafolder["Real Blade"]))  -- Outputs "true" and remaining currency if purchase succeeds

Reviewer: ```suggestion
```lua title="LocalScript in StarterPlayerScripts" highlight="5, 17, 20"
local ReplicatedStorage = game:GetService("ReplicatedStorage")

local itemDataFolder = ReplicatedStorage:WaitForChild("ItemData")
local buyItemEvent = ReplicatedStorage:WaitForChild("BuyItemEvent")
local payload = {
	Name = "Ultra Blade",
	ClassName = "Folder",
	Parent = itemDataFolder,
	Price = {
		Name = "Price",
		ClassName = "NumberValue",
		Value = 0,  -- Negative values could also be used, resulting in giving currency rather than taking it!
	},
}

-- Send malicious payload to the server (this will be rejected)
print(buyItemEvent:InvokeServer(payload))  -- Outputs "false Invalid item provided"

-- Send a real item to the server (this will go through!)
print(buyItemEvent:InvokeServer(itemDatafolder["Real Blade"]))  -- Outputs "true" and remaining currency if purchase succeeds
```
```

```lua title="Script in ServerScriptService" highlight="7-10"
local ReplicatedStorage = game:GetService("ReplicatedStorage")

local itemDataFolder = ReplicatedStorage:WaitForChild("ItemData")
local buyItemEvent = ReplicatedStorage:WaitForChild("BuyItemEvent")

> buyItemEvent.OnServerInvoke = function(player, item)
> 	-- Check if the passed item isn't spoofed and is in the ItemData folder
> 	if typeof(item) ~= "Instance" or not item:IsDescendantOf(itemDataFolder) then
> 		return false, "Invalid item provided"
> 	end
> 
> 	-- The server can then go on to process the purchase based on the example flow below
> end

Reviewer: ```suggestion
local function buyItem(player, item)
	-- Check if the passed item isn't spoofed and is in the ItemData folder
	if typeof(item) ~= "Instance" or not item:IsDescendantOf(itemDataFolder) then
		return false, "Invalid item provided"
	end

	-- The server can then go on to process the purchase based on the example flow below
end

-- Bind "buyItem()" to the remote function's callback
buyItemEvent.OnServerInvoke = buyItem
```

Reviewer: Basically just declaring a dedicated non-anonymous function to fit style of example above.

Author: Very good change! Allows for usage from other potential avenues of purchasing (bulk purchases etc.)

end
```

Another common attack that exploiters may use involves sending `Library.table|tables` in place of an `Class.Instance`. Complex payloads can mimic what would be an otherwise ordinary object reference.

For example, provided with an [in-experience shop](#in-experience-shop) system where item data like prices are stored in `Class.NumberValue` objects, an exploiter may circumvent all other checks by doing the following:

> ```lua title="LocalScript in StarterPlayerScripts" highlight="5, 17"
> local ReplicatedStorage = game:GetService("ReplicatedStorage")
> 
> local itemDataFolder = ReplicatedStorage:WaitForChild("ItemData")
> local buyItemEvent = ReplicatedStorage:WaitForChild("BuyItemEvent")
> local payload = {
> 	Name = "Ultra Blade",
> 	ClassName = "Folder",
> 	Parent = itemDataFolder,
> 
> 	Price = {
> 		Name = "Price",
> 		ClassName = "NumberValue",
> 		Value = 0, -- Negative values could also be used which would result in giving currency rather than taking it!
> 	},
> }
> 
> -- Send malicious payload to the server. This will be rejected
> print(buyItemEvent:InvokeServer(payload))
> -- ^ Outputs "false Invalid item provided"
> 
> -- Send a real item to the server. This will go through!
> print(buyItemEvent:InvokeServer(itemDatafolder["Real Blade"]))
> -- ^ Outputs "true" and remaining currency if purchase succeeds
> ```

Reviewer: ```suggestion
```lua title="LocalScript in StarterPlayerScripts" highlight="5, 18, 21"
local ReplicatedStorage = game:GetService("ReplicatedStorage")

local itemDataFolder = ReplicatedStorage:WaitForChild("ItemData")
local buyItemEvent = ReplicatedStorage:WaitForChild("BuyItemEvent")
local payload = {
	Name = "Ultra Blade",
	ClassName = "Folder",
	Parent = itemDataFolder,

	Price = {
		Name = "Price",
		ClassName = "NumberValue",
		Value = 0,  -- Negative values could also be used, resulting in giving currency rather than taking it!
	},
}

-- Send malicious payload to the server (this will be rejected)
print(buyItemEvent:InvokeServer(payload))  -- Outputs "false Invalid item provided"

-- Send a real item to the server (this will go through!)
print(buyItemEvent:InvokeServer(itemDatafolder["Real Blade"]))  -- Outputs "true" and remaining currency if purchase succeeds
```
```

- Confirm that the hit player is alive.
- Store weapon and player state on the server and confirm that a firing player is not blocked by a current action such as reloading or a state like sprinting.

#### DataStore Manipulation

> In games using `Class.DataStoreService` to save player data, exploiters may take advantage of invalid [data](#data-validation), and more obscure methods, to prevent a DataStore from saving properly. This can be especially abused in games with item trading, marketplaces and similar systems where items or currency leave a player's inventory.
> 
> Ensure that any actions performed through a `Class.RemoteEvent` or `Class.RemoteFunction` that affect player data with client input is sanitised based on the following:
> 
> - `Class.Instance` values cannot be serialised into a DataStore and will fail. Utilise [type validation](#remote-runtime-type-validation) to prevent this.

Reviewer: ```suggestion
In experiences using `Class.DataStoreService` to save player data, exploiters may take advantage of invalid [data](#data-validation), and more obscure methods, to prevent a `Class.DataStore` from saving properly. This can be especially abused in experiences with item trading, marketplaces, and similar systems where items or currency leave a player's inventory.

Ensure that any actions performed through a `Class.RemoteEvent` or `Class.RemoteFunction` that affect player data with client input is sanitized based on the following:

- `Class.Instance` values cannot be serialized into a `Class.DataStore` and will fail. Utilize [type validation](#remote-runtime-type-validation) to prevent this.
```

> - DataStores have a data size limit per key of 4 megabytes. Strings of arbitrary length should be checked and/or capped to avoid this, alongside ensuring limitless arbitrary keys cannot be added to tables by the client.

Reviewer: ```suggestion
- `Class.DataStore|DataStores` have a data [limits](../../cloud-services/datastores#limits). Strings of arbitrary length should be checked and/or capped to avoid this, alongside ensuring limitless arbitrary keys cannot be added to tables by the client.
```

Reviewer: Linking to "Limits" section of Data Stores article (https://create.roblox.com/docs/cloud-services/datastores), in case the limit of 4 MB changes in the future and effectively makes this citation outdated.

Author: This is true. I didn't notice this when writing (probably due to how rarely these limits are changed LOL). Good catch!

> - Table indices cannot be [NaN](#data-validation) or nil. Iterate over all tables passed by the client and verify all indices are valid.
> - DataStores can only accept valid UTF-8 characters. Sanitise all strings provided by the client to ensure only bytes 1-127 are used. By using `Library.string.find()` with the pattern `"[%z\128-\255]"` an operation can be aborted if any invalid characters are found in a string.

Reviewer: ```suggestion
- Table indices cannot be `NaN` or `nil`. Iterate over all tables passed by the client and verify all indices are valid.
- `Class.DataStore|DataStores` can only accept valid UTF-8 characters. Sanitize all strings provided by the client to ensure only bytes 1-127 are used. By using `Library.string.find()` with the pattern `"[%z\128-\255]"`, an operation can be aborted if any invalid characters are found in a string.
```

Author: I linked back to data-validation here as it provides functions to check for NaN. If someone jumps to this point of the article there's a chance they could miss this function. That's pretty unlikely though, so this change is fine.

- Confirm that the hit player is alive.
- Store weapon and player state on the server and confirm that a firing player is not blocked by a current action such as reloading or a state like sprinting.

#### DataStore Manipulation

In games using `Class.DataStoreService` to save player data, exploiters may take advantage of invalid [data](#data-validation), and more obscure methods, to prevent a DataStore from saving properly. This can be especially abused in games with item trading, marketplaces and similar systems where items or currency leave a player's inventory.

Ensure that any actions performed through a `Class.RemoteEvent` or `Class.RemoteFunction` that affect player data with client input is sanitised based on the following:

> - `Class.Instance` values cannot be serialised into a DataStore and will fail. Utilise [type validation](#remote-runtime-type-validation) to prevent this.

Author: Issue fixed.

- Confirm that the hit player is alive.
- Store weapon and player state on the server and confirm that a firing player is not blocked by a current action such as reloading or a state like sprinting.

#### DataStore Manipulation

In experiences using `Class.DataStoreService` to save player data, exploiters may take advantage of invalid [data](#data-validation), and more obscure methods, to prevent a `Class.DataStore` from saving properly. This can be especially abused in experiences with item trading, marketplaces, and similar systems where items or currency leave a player's inventory.

Ensure that any actions performed through a `Class.RemoteEvent` or `Class.RemoteFunction` that affect player data with client input is sanitized based on the following:

- `Class.Instance` values cannot be serialized into a `Class.DataStore` and will fail. Utilize [type validation](#remote-runtime-type-validation) to prevent this.

> - `Class.DataStore|DataStores` have data [limits](../../cloud-services/datastores#limits). Strings of arbitrary length should be checked and/or capped to avoid this, alongside ensuring limitless arbitrary keys cannot be added to tables by the client.

Author: This is fixed now. The current content/quality checks are failing for a completely different file for some reason. Completely out of scope of this article (data-store-api-handling.md isn't even referenced in this file!)
